### PR TITLE
RTH: Wait until descend speed has been over 25cm/s before starting to detect landing

### PR DESCRIPTION
**Pull request by theArchLadder**
As discussed in #219 
Work in progress, not ready for merging!

This have been flight tested and it looks like it works great.
hasHadSomeVelocity and !verticalMovement are never true at the same time during the beginning of the descent, this should make things safer, reduce the need for a long timeout and make having a to high min_fly setting less dangerous.

Thoughts, ideas for improvement and other feedback are welcome!

EDIT: I also removed the code that only relied on the sonar, that's to dangerous. The commit named "Fixed landingTimer reset" is about hasHadSomeVelocity, i have not fixed the bug i talked about in the issue, sorry for making a bit of a mess here.

### CODE REVIEW

**Comment by theArchLadder** [comment]
The last commit has not been flight tested yet, but it did build without warnings.

**Comment by digitalentity** [comment]
I think we should also make sure that target descent velocity is always above 25cm/s here: https://github.com/iNavFlight/inav/blob/master/src/main/flight/navigation_rewrite.c#L959

Or maybe even above 30-40 cm/s to be certain.

File: src/main/flight/navigation_rewrite_multicopter.c [hunk 1]
```diff
@@ -483,11 +486,19 @@ bool isMulticopterLandingDetected(uint32_t * landingTimer)
     // from processRx() and rcCommand at that moment holds rc input, not adjusted values from NAV core)
     bool minimalThrust = rcCommandAdjustedThrottle < posControl.navConfig->mc_min_fly_throttle;
 
+    /*
+    Do not trust the sonar, it wants to kill you!
```
Comment (digitalentity) [inline src/main/flight/navigation_rewrite_multicopter.c]: Sonar is ok, we just need a better way to detect if its healthy and can be trusted.

Comment (theArchLadder) [inline src/main/flight/navigation_rewrite_multicopter.c]: It was not ok in the blackbox logs posted in the issue, have you had a chance to look at them? Urass sonar reads 4-6cm very stable while the sonar is out of range, and in stronnag's logs the reading was very noisy.
I guess a noisy sonar is easy to detect as unhealthy, but what about that stable out of range reading that looks exactly like it's sitting on the ground...?

I think it's a bad idea to only trust the sonar, but as a added safety, sure!

**Comment by theArchLadder** [comment]
> I think we should also make sure that target descent velocity is always above 25cm/s here: 

Do you mean we should check if it is above, or hard code it so that it never can be below 30-40cm/s...?

**Comment by digitalentity** [comment]
I ment hardcode the lowest possible velocity to make sure we never go
slower unless on the ground.

**Comment by theArchLadder** [comment]
I agree, i had an idea of using `posControl.navConfig->land_descent_rate * 0.125f` instead of `25.0f` in the landing detection code, this would allow the user to reduce touchdown velocity but if one uses really low value the `verticalMovement` test may have trouble being passed for 2 seconds in a row...

So, yeah, i think your idea is better and more robust, 50cm/s seems really slow and gentle for touchdown anyway.

I did a test flight again now but with only a 200ms timeout (so it has to pass 10 times in a row), and i set my min_fly only around 70 below my hover throttle.
It disarmed almost the same instant as it touched down, looked very cool! I think i did more than 10 landings and i didn't have a single "disarm in air" even with these risky settings! The nice thing is that `(hasHadSomeVelocity && !verticalMovement)` never passes until the quad is on the ground, so even with min_fly wrongly set over hover throttle, it would still work :+1: 
I don't suggest using these as default, but pushing it to the limit is a good and fun way to test the robustness of the landing detection imho.

BTW: Did you read this?
https://github.com/iNavFlight/inav/issues/219#issuecomment-220793432
Not sure if we should do it all in this pull request or if it would be better to keep this pull focused on >25cm/s stuff and merge it in soon, how would you like it? I don't know how to solve that bug.

**Comment by theArchLadder** [comment]
Last commits flight tested and works as expected.
With a constant -50cm/s descent velocity, my slowest navVel[2] was -28cm/s, so i feel that a slowest allowed target descent velocity of -50cm/s is about right.

I also changed the sonar code so it's only used as a extra safety for now, since both urass and stronnag had "disarm in air" with the old code, we can do something more clever when we have a healthy-sonar-check working. Github hides the code-comments so press "show 2 commets" above here if you didn't read my response.

**Comment by DzikuVx** [comment]
I have an idea of fools proof landing detection, but slightly more complicated that current one.
Idea of detecting landing by measuring only vertical speed has a problem: vertical speed depends on many factors like air cushion, wind gusts and so on. For sonar it is even such a detail like type of surface (grass, concrete). I can easily imagine at least few reasons for which vertical speed can be reduced below any reasonable threshold a meter above ground.
Even checking for `mc_min_fly_throttle` has a "user" problem. Pilot has to know it, has to set it and in 90% of the cases he/she will not do that.
Idea is to have additional _active_ check. When current detection returns `true` try cutting throttle by `x` for `time` and check accelerometer response. If UAV starts to accelerate, in any axis, ground has not been touched, we have false positive, abort. If accelerometer did not detected anything for `time` landing has been completed.
What do you think?

**Comment by theArchLadder** [comment]
@DzikuVx I have also been thinking about using the accelerometer!
As an "false positive"-detector it sounds clever, but i think we still want it to work without it, trail and error landing detection could look a bit strange "cut throttle... whops, we are falling, increase throttle again! phew, saved it... okay lets repeat"
Having it as an extra safety feature would allow us to take more risk with the current logic, for example we could reduce the timeout so we disarm faster, that's still a big improvement!

As for this PR, i think i'm about done with what i wanted to do here, how would you guys like to continue? I find those debug values useful for learning more about this, would be great to see more bblogs from different copters... Perhaps make a testing branch for this?

**Comment by digitalentity** [comment]
@DzikuVx sounds like a good idea to me. Another way is not to cut throttle but modulate it and see if accelerometer follows - if it does - we are still flying.

**Comment by digitalentity** [comment]
@theArchLadder do you think its safe to merge it now?

**Comment by theArchLadder** [comment]
Merge it into master? In that case would you like me to remove or comment out the navDebug[]?
Besides that i think it's ready!

**Comment by digitalentity** [comment]
@theArchLadder yes, pelase cleanup and I'll merge it

**Comment by theArchLadder** [comment]
Done!

**Pull request merged**